org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/plain/UpdateProjectClasspathCommand.java

```diff
@@ -80,23 +84,47 @@ public static Object execute(List<Object> arguments, IProgressMonitor pm) {
   private static IClasspathEntry[] createModifiedEntry(List<ClasspathEntry> entries) {
     List<IClasspathEntry> coreClasspathEntries = new ArrayList<>(entries.size());
     for (ClasspathEntry entry : entries) {
-      IPath path = fromOSString(entry.getPath());
-      switch (entry.getEntryKind()) {
-        case CPE_LIBRARY:
-          coreClasspathEntries.add(newLibraryEntry(path, null, null));
-          break;
-        case CPE_SOURCE:
-          coreClasspathEntries.add(newSourceEntry(path));
-          break;
-        case CPE_VARIABLE:
-          coreClasspathEntries.add(newVariableEntry(path, null, null));
-          break;
-        case CPE_CONTAINER:
-          coreClasspathEntries.add(newContainerEntry(path));
-          break;
-        case CPE_PROJECT:
-          coreClasspathEntries.add(newProjectEntry(path));
-          break;
+      try {
+        switch (entry.getEntryKind()) {
+          case CPE_LIBRARY:
+            {
+              String absolutePath = new File(new URI(entry.getPath())).getAbsolutePath();
+              coreClasspathEntries.add(
+                  newLibraryEntry(Path.fromOSString(absolutePath), null, null));
+              break;
+            }
+          case CPE_SOURCE:
+            {
+              IContainer[] folders =
+                  ResourcesPlugin.getWorkspace()
+                      .getRoot()
+                      .findContainersForLocationURI(new URI(entry.getPath()));
+              if (folders.length == 0) {
+                throw new IllegalArgumentException("no folders found for " + entry.getPath());
+              }
+              coreClasspathEntries.add(newSourceEntry(folders[0].getFullPath()));
+              break;
+            }
+          case CPE_VARIABLE:
+            coreClasspathEntries.add(
+                newVariableEntry(Path.fromOSString(entry.getPath()), null, null));
+            break;
+          case CPE_CONTAINER:
+            coreClasspathEntries.add(newContainerEntry(Path.fromOSString(entry.getPath())));
+            break;
+          case CPE_PROJECT:
+            IContainer[] folders =
+                ResourcesPlugin.getWorkspace()
+                    .getRoot()
+                    .findContainersForLocationURI(new URI(entry.getPath()));
+            if (folders.length == 0) {
+              throw new IllegalArgumentException("no folders found for " + entry.getPath());
+            }
```

Comment:
it would be better to move code for finding container by uri into separate method and then use it for `CPE_CONTAINER` and `CPE_PROJECT`.